### PR TITLE
VPA Recommender Sink and Recommender JSON Client added.

### PR DESCRIPTION
**Pull request by KarolKraskiewicz**
This is a DataSink implementation, which will be sending container utilization data to VPA Recommender, described in [Pull Request 338](https://github.com/kubernetes/community/pull/338)

@mwielgus @kgrygiel and @piosz please review.

### CODE REVIEW

**Comment by KarolKraskiewicz** [comment]
I signed it! (The cla)

**Comment by KarolKraskiewicz** [comment]
I decided to push 1 updated to Recommender, consolidating info about all containers, instead of sending single request per container. It should save some time due to network latency and simplify processing on Recommender side. 

It means however, we will be able to send information about only up to 20 000 containers:
- info about single container = 512B
- 512B x 20 000 = 10MB which is a limit of body size when handling http post request in go.

The 20k containers limit will be raised when sink/recommender communication protocol is changed from JSON to Protocol Buffer. It should get us to up to 30-40k and from here we can go further by switching to multipart post.

**Comment by DirectXMan12** [comment]
Is this intended to be a temporary solution, or a long-term one?  In the long term, I don't think we should tie the VPA to Heapster, since Heapster will eventually no longer be a necessary cluster component.

Is there any reason the VPA can't pull this information from the resource metrics API instead?  That way, everything will continue to work if people run metrics-server or something else.

File: metrics/core/types.go [hunk 1]
```diff
@@ -117,14 +117,20 @@ func (this *LabeledMetric) GetValue() interface{} {
 	}
 }
 
+//all the metrics from single resource (like pod or container)
 type MetricSet struct {
-	CreateTime     time.Time
```
Comment (mwielgus) [inline metrics/core/types.go]: Let's move these comments to a separate pr.

File: metrics/sinks/vpa/container_utilization.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+type containerUtilizationSnapshot struct {
+	CreateTime     time.Time `json:"createTime"`
+	ScrapTime      time.Time `json:"scrapTime"`
+	ContainerName  string    `json:"containerName"`
+	ContainerImage string    `json:"containerImage"`
+	PodId          string    `json:"podId"`
+	// Amount of requested memory.
+	// Units: Bytes.
+	MemoryRequest int64 `json:"memoryRequest"`
```
Comment (mwielgus) [inline metrics/sinks/vpa/container_utilization.go]: I guess we should probably use maps here. CPU and Memory are not the only resources we may want to use.

File: metrics/sinks/vpa/container_utilization.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+func (set *utilizationSnapshotSet) addContainer(utilizationSnapshot *containerUtilizationSnapshot) {
+	set.Containers = append(set.Containers, utilizationSnapshot)
+}
+
+type containerUtilizationSnapshot struct {
+	CreateTime     time.Time `json:"createTime"`
+	ScrapTime      time.Time `json:"scrapTime"`
+	ContainerName  string    `json:"containerName"`
+	ContainerImage string    `json:"containerImage"`
+	PodId          string    `json:"podId"`
```
Comment (mwielgus) [inline metrics/sinks/vpa/container_utilization.go]: PodID (as uid) is useless. Replace with podname + podnamespace.

File: metrics/sinks/vpa/container_utilization.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+package vpa
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/heapster/metrics/core"
+	"time"
+)
+
+type utilizationSnapshotSet struct {
```
Comment (mwielgus) [inline metrics/sinks/vpa/container_utilization.go]: UpperCase

Comment (mwielgus) [inline metrics/sinks/vpa/container_utilization.go]: Add an information that this set may be only a pice of a multipart batch.

File: metrics/sinks/vpa/recommender.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+			glog.Warningf("Unable to create utilization snapshot, due to error: %s", err.Error())
+		}
+	}
+
+	sink.sendAllSnapshots(&snapshotSet)
+}
+
+func (sink *recommenderSink) sendAllSnapshots(set *utilizationSnapshotSet) {
+	if len(set.Containers) > 0 {
+		_, err := sink.recommenderClient.SendJSON(set)
```
Comment (mwielgus) [inline metrics/sinks/vpa/recommender.go]: Add batching.

File: metrics/sinks/vpa/recommender.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+func (sink *recommenderSink) Name() string {
+	return "VPA Recommender Sink"
+}
+
+func (sink *recommenderSink) Stop() {
+	// nothing needs to be done.
+}
+
+func isContainerMetricSet(metricSet *core.MetricSet) bool {
+	return metricSet.Labels[core.LabelContainerName.Key] != "" && metricSet.Labels[core.LabelContainerBaseImage.Key] != ""
```
Comment (mwielgus) [inline metrics/sinks/vpa/recommender.go]: Please check if we have a helper function for this.

**Comment by mwielgus** [comment]
@DirectXMan12 
Yeah, unfortunately I have to agree with you :). We had a conversation with @piosz and @KarolKraskiewicz, @GFilipek and decided that reversing the data traffic makes sense.

**Comment by DirectXMan12** [comment]
We should close this, then
